test/main.lua

```diff
@@ -292,55 +379,75 @@ function run_test(test, status, counts)
-        file_errors = 0,
-    }
-    local passed = true
-
-    ensure_title_screen()
-
-    print('Loading tests')
+local function get_tests(test_files, counts)
     local tests = {}
-    for _, file in ipairs(files) do
+    for _, file in ipairs(test_files) do
         if not load_tests(file, tests) then
-            passed = false
             counts.file_errors = counts.file_errors + 1
         end
     end
+    return tests
+end
 
-    print('Filtering tests')
-    if config.tests then
+local function filter_tests(tests, config)
+    if config.tests or config.modes then
+        for _,filter in ipairs({{'tests', 'name pattern'},{'modes', 'mode'}}) do
+            if config[filter[1]] and #config[filter[1]] > 0 then
+                print(string.format('Filtering tests by %s:', filter[2]))
+                for _,v in ipairs(config[filter[1]]) do
+                    print(string.format('  %s', v))
+                end
+            end
+        end
         local orig_length = #tests
         for i = #tests, 1, -1 do
-            for _, pattern in pairs(config.tests) do
-                if not tests[i].name:match(pattern) then
-                    table.remove(tests, i)
+            local remove = false
+            if config.modes then
+                remove = true
+                -- allow test if it matches any of the given modes
+                for _, mode in pairs(config.modes) do
+                    if tests[i].config.mode == mode then
+                        remove = false
+                        break
+                    end
                 end
             end
+            if config.tests and not remove then
+                remove = true
+                -- allow test if it matches any of the given patterns
+                for _, pattern in pairs(config.tests) do
+                    if tests[i].name:match(pattern) then
+                        remove = false
+                        break
+                    end
+                end
+            end
+            if remove then table.remove(tests, i) end
         end
         print('Selected tests: ' .. #tests .. '/' .. orig_length)
     end
-    local status = load_test_status() or {}
-    for i = #tests, 1, -1 do
-        local test = tests[i]
-        if not status[test.full_name] then
-            status[test.full_name] = TestStatus.PENDING
-        elseif status[test.full_name] ~= TestStatus.PENDING then
-            print('skipping test: ' .. test.name .. ': state = ' .. status[test.full_name] .. ')')
-            table.remove(tests, i)
+
+    local status = {}
+    if not config.nocache then
+        status = load_test_status() or status
+        for i = #tests, 1, -1 do
+            local test = tests[i]
+            if not status[test.full_name] then
+                status[test.full_name] = TestStatus.PENDING
+            elseif status[test.full_name] ~= TestStatus.PENDING then
+                print(('skipping test: %s: state = %s)'):format(
```

Comment:
Heh, thanks for preserving the `)`, but that's a typo from e6c9db0a18a49ab3777d904151a40bb1d660a398 :)

Reply:
haha, I noticed, but I thought it was just stylized formatting -- it actually looks reasonable in the output